base/abstractarray.jl

```diff
@@ -440,13 +440,15 @@ false
 checkindex(::Type{Bool}, inds::AbstractUnitRange, i) = throw(ArgumentError("unable to check bounds for indices of type $(typeof(i))"))
 checkindex(::Type{Bool}, inds::AbstractUnitRange, i::Real) = (first(inds) <= i) & (i <= last(inds))
 checkindex(::Type{Bool}, inds::AbstractUnitRange, ::Colon) = true
-checkindex(::Type{Bool}, inds::AbstractUnitRange, ::Slice) = true
+checkindex(::Type{Bool}, inds::AbstractUnitRange, s::Slice) =
+    (first(s) >= first(inds)) & (last(s) <= last(inds))
```

Comment:
I think that'd put us right back to where we are now, only `s/Slice/Colon/`.

In order for Colon to be a range, it must know where to start and how long it is.  That can only be resolved upon indexing (within `to_indices`).  Thus, there must be some way of constructing a colon with a given start and length.  But if a user ends up using that same construction then it might not accurately describe an entire dimension.  And now we're right back to where we started, except this time it's a more visible, exported object that users regularly construct.